[ICE]: vtable : `None` from `prepare_vtable_segments()`

<!--
[31mICE[0m: Rustc ./4D3F8D4D46E1245F68AAAD1F329B376BA3166264E585638CB073374363186000.rs '-Znext-solver=globally -ooutputfile -Zdump-mir-dir=dir' 'thread 'rustc' ($TID) panicked at compiler/rustc_trait_selection/src/traits/vtable.rs:437:77: 'called `Option::unwrap()` on a `None` value'', 'thread 'rustc' ($TID) panicked at compiler/rustc_trait_selection/src/traits/vtable.rs:437:77: 'called `Option::unwrap()` on a `None` value''
File: /tmp/im/4D3F8D4D46E1245F68AAAD1F329B376BA3166264E585638CB073374363186000.rs
-->

auto-reduced (treereduce-rust):
````rust
//@compile-flags: -Znext-solver=globally
trait Supertrait<T> {}

trait Trait<P>: Supertrait {}

impl<P> Trait<P> for () {}

const fn upcast<P>(x: &dyn Trait<P>) -> &dyn Trait<P> {
    x
}

const fn foo() -> &'static dyn Supertrait<()> {
    upcast::<()>(&())
}

const _: &'static dyn Supertrait<()> = foo();

````

original:
````rust
trait Supertrait<T> {
    fn static(&self) {}
}

trait Trait<P>: Supertrait<Output = impl ReturnsSend> {}

impl<P> Trait<P> for () {}

const fn upcast<P>(x: &dyn Trait<P>) -> &dyn Trait<P> {
    x
}

const fn foo() -> &'static dyn Supertrait<()> {
    upcast::<()>(&())
}

const _: &'static dyn Supertrait<()> = foo();

````

Version information
````
rustc 1.95.0-nightly (621d76794 2026-01-18)
binary: rustc
commit-hash: 621d76794c76fc21c0a6151fbc090120e0230a91
commit-date: 2026-01-18
host: x86_64-unknown-linux-gnu
release: 1.95.0-nightly
LLVM version: 21.1.8

````

Possibly related line of code:
https://github.com/rust-lang/rust/blob/621d76794c76fc21c0a6151fbc090120e0230a91/compiler/rustc_trait_selection/src/traits/vtable.rs#L431-L443

Command:
`/home/matthias/.rustup/toolchains/master/bin/rustc -Znext-solver=globally`

<details><summary><strong>Program output</strong></summary>
<p>

```
error[E0601]: `main` function not found in crate `mvce`
  --> /tmp/icemaker_global_tempdir.K5SjLYH7a8mL/rustc_testrunner_tmpdir_reporting.EuEghOYFieqJ/mvce.rs:15:46
   |
15 | const _: &'static dyn Supertrait<()> = foo();
   |                                              ^ consider adding a `main` function to `/tmp/icemaker_global_tempdir.K5SjLYH7a8mL/rustc_testrunner_tmpdir_reporting.EuEghOYFieqJ/mvce.rs`

error[E0107]: missing generics for trait `Supertrait`
 --> /tmp/icemaker_global_tempdir.K5SjLYH7a8mL/rustc_testrunner_tmpdir_reporting.EuEghOYFieqJ/mvce.rs:3:17
  |
3 | trait Trait<P>: Supertrait {}
  |                 ^^^^^^^^^^ expected 1 generic argument
  |
note: trait defined here, with 1 generic parameter: `T`
 --> /tmp/icemaker_global_tempdir.K5SjLYH7a8mL/rustc_testrunner_tmpdir_reporting.EuEghOYFieqJ/mvce.rs:1:7
  |
1 | trait Supertrait<T> {}
  |       ^^^^^^^^^^ -
help: add missing generic argument
  |
3 | trait Trait<P>: Supertrait<T> {}
  |                           +++


thread 'rustc' (2079460) panicked at /rustc-dev/621d76794c76fc21c0a6151fbc090120e0230a91/compiler/rustc_trait_selection/src/traits/vtable.rs:437:77:
called `Option::unwrap()` on a `None` value
stack backtrace:
   0:     0x7f354f433fd3 - <<std[287527946f71f0bc]::sys::backtrace::BacktraceLock>::print::DisplayBacktrace as core[7477a31a1a34cc49]::fmt::Display>::fmt
   1:     0x7f354fa0e2c8 - core[7477a31a1a34cc49]::fmt::write
   2:     0x7f354f44a9c6 - <std[287527946f71f0bc]::sys::stdio::unix::Stderr as std[287527946f71f0bc]::io::Write>::write_fmt
   3:     0x7f354f40a048 - std[287527946f71f0bc]::panicking::default_hook::{closure#0}
   4:     0x7f354f427703 - std[287527946f71f0bc]::panicking::default_hook
   5:     0x7f354e3fbb8a - std[287527946f71f0bc]::panicking::update_hook::<alloc[979041e9952e771b]::boxed::Box<rustc_driver_impl[517260a07841a630]::install_ice_hook::{closure#1}>>::{closure#0}
   6:     0x7f354f4279e2 - std[287527946f71f0bc]::panicking::panic_with_hook
   7:     0x7f354f40a13a - std[287527946f71f0bc]::panicking::panic_handler::{closure#0}
   8:     0x7f354f400dc9 - std[287527946f71f0bc]::sys::backtrace::__rust_end_short_backtrace::<std[287527946f71f0bc]::panicking::panic_handler::{closure#0}, !>
   9:     0x7f354f40bbdd - __rustc[5ec7081ce87b032f]::rust_begin_unwind
  10:     0x7f354bf0cfbc - core[7477a31a1a34cc49]::panicking::panic_fmt
  11:     0x7f354bbe10a2 - core[7477a31a1a34cc49]::panicking::panic
  12:     0x7f354c27c749 - core[7477a31a1a34cc49]::option::unwrap_failed
  13:     0x7f354f259c40 - rustc_trait_selection[9fee51d8a10a777c]::traits::vtable::supertrait_vtable_slot
  14:     0x7f354ee153e6 - rustc_query_impl[f20b8284200e680a]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[f20b8284200e680a]::query_impl::supertrait_vtable_slot::dynamic_query::{closure#2}::{closure#0}, rustc_middle[725063e074f29df]::query::erase::Erased<[u8; 16usize]>>
  15:     0x7f354ef17f59 - <rustc_query_impl[f20b8284200e680a]::query_impl::supertrait_vtable_slot::dynamic_query::{closure#2} as core[7477a31a1a34cc49]::ops::function::FnOnce<(rustc_middle[725063e074f29df]::ty::context::TyCtxt, (rustc_middle[725063e074f29df]::ty::Ty, rustc_middle[725063e074f29df]::ty::Ty))>>::call_once
  16:     0x7f354ee28596 - rustc_query_system[b7e73caa2c2ad9a]::query::plumbing::try_execute_query::<rustc_query_impl[f20b8284200e680a]::DynamicConfig<rustc_query_system[b7e73caa2c2ad9a]::query::caches::DefaultCache<(rustc_middle[725063e074f29df]::ty::Ty, rustc_middle[725063e074f29df]::ty::Ty), rustc_middle[725063e074f29df]::query::erase::Erased<[u8; 16usize]>>, false, false, false>, rustc_query_impl[f20b8284200e680a]::plumbing::QueryCtxt, false>
  17:     0x7f354eea1340 - rustc_query_impl[f20b8284200e680a]::query_impl::supertrait_vtable_slot::get_query_non_incr::__rust_end_short_backtrace
  18:     0x7f354ff9e418 - <rustc_const_eval[b948c5b36ae67fb9]::interpret::eval_context::InterpCx<rustc_const_eval[b948c5b36ae67fb9]::const_eval::machine::CompileTimeMachine>>::unsize_into_ptr::<rustc_const_eval[b948c5b36ae67fb9]::interpret::place::PlaceTy>
  19:     0x7f3550b00904 - <rustc_const_eval[b948c5b36ae67fb9]::interpret::eval_context::InterpCx<rustc_const_eval[b948c5b36ae67fb9]::const_eval::machine::CompileTimeMachine>>::eval_rvalue_into_place
  20:     0x7f3550aba0cd - rustc_const_eval[b948c5b36ae67fb9]::const_eval::eval_queries::eval_to_allocation_raw_provider
  21:     0x7f3550ab847e - rustc_query_impl[f20b8284200e680a]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[f20b8284200e680a]::query_impl::eval_to_allocation_raw::dynamic_query::{closure#2}::{closure#0}, rustc_middle[725063e074f29df]::query::erase::Erased<[u8; 24usize]>>
  22:     0x7f3550ab7ba5 - rustc_query_system[b7e73caa2c2ad9a]::query::plumbing::try_execute_query::<rustc_query_impl[f20b8284200e680a]::DynamicConfig<rustc_query_system[b7e73caa2c2ad9a]::query::caches::DefaultCache<rustc_middle[725063e074f29df]::ty::PseudoCanonicalInput<rustc_middle[725063e074f29df]::mir::interpret::GlobalId>, rustc_middle[725063e074f29df]::query::erase::Erased<[u8; 24usize]>>, false, false, false>, rustc_query_impl[f20b8284200e680a]::plumbing::QueryCtxt, false>
  23:     0x7f3550ab7749 - rustc_query_impl[f20b8284200e680a]::query_impl::eval_to_allocation_raw::get_query_non_incr::__rust_end_short_backtrace
  24:     0x7f3550ab3504 - rustc_const_eval[b948c5b36ae67fb9]::const_eval::eval_queries::eval_to_const_value_raw_provider
  25:     0x7f3550ab3080 - rustc_query_impl[f20b8284200e680a]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[f20b8284200e680a]::query_impl::eval_to_const_value_raw::dynamic_query::{closure#2}::{closure#0}, rustc_middle[725063e074f29df]::query::erase::Erased<[u8; 24usize]>>
  26:     0x7f3550ab1c8a - rustc_query_system[b7e73caa2c2ad9a]::query::plumbing::try_execute_query::<rustc_query_impl[f20b8284200e680a]::DynamicConfig<rustc_query_system[b7e73caa2c2ad9a]::query::caches::DefaultCache<rustc_middle[725063e074f29df]::ty::PseudoCanonicalInput<rustc_middle[725063e074f29df]::mir::interpret::GlobalId>, rustc_middle[725063e074f29df]::query::erase::Erased<[u8; 24usize]>>, false, true, false>, rustc_query_impl[f20b8284200e680a]::plumbing::QueryCtxt, false>
  27:     0x7f3550ab17f3 - rustc_query_impl[f20b8284200e680a]::query_impl::eval_to_const_value_raw::get_query_non_incr::__rust_end_short_backtrace
  28:     0x7f35505401b0 - <rustc_middle[725063e074f29df]::ty::context::TyCtxt>::par_hir_body_owners::<rustc_hir_analysis[1a80394817eddfdd]::check_crate::{closure#2}>::{closure#0}
  29:     0x7f355053f7c4 - rustc_hir_analysis[1a80394817eddfdd]::check_crate
  30:     0x7f354fb493db - rustc_interface[5c48368f5debe80d]::passes::analysis
  31:     0x7f354fb49035 - rustc_query_impl[f20b8284200e680a]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[f20b8284200e680a]::query_impl::analysis::dynamic_query::{closure#2}::{closure#0}, rustc_middle[725063e074f29df]::query::erase::Erased<[u8; 0usize]>>
  32:     0x7f3550ba4be3 - rustc_query_system[b7e73caa2c2ad9a]::query::plumbing::try_execute_query::<rustc_query_impl[f20b8284200e680a]::DynamicConfig<rustc_query_system[b7e73caa2c2ad9a]::query::caches::SingleCache<rustc_middle[725063e074f29df]::query::erase::Erased<[u8; 0usize]>>, false, false, false>, rustc_query_impl[f20b8284200e680a]::plumbing::QueryCtxt, false>
  33:     0x7f3550ba49ce - rustc_query_impl[f20b8284200e680a]::query_impl::analysis::get_query_non_incr::__rust_end_short_backtrace
  34:     0x7f3550d82c11 - <rustc_interface[5c48368f5debe80d]::passes::create_and_enter_global_ctxt<core[7477a31a1a34cc49]::option::Option<rustc_interface[5c48368f5debe80d]::queries::Linker>, rustc_driver_impl[517260a07841a630]::run_compiler::{closure#0}::{closure#2}>::{closure#2} as core[7477a31a1a34cc49]::ops::function::FnOnce<(&rustc_session[eb79e6d3104f29ae]::session::Session, rustc_middle[725063e074f29df]::ty::context::CurrentGcx, alloc[979041e9952e771b]::sync::Arc<rustc_data_structures[e0d8483206daafa2]::jobserver::Proxy>, &std[287527946f71f0bc]::sync::once_lock::OnceLock<rustc_middle[725063e074f29df]::ty::context::GlobalCtxt>, &rustc_data_structures[e0d8483206daafa2]::sync::worker_local::WorkerLocal<rustc_middle[725063e074f29df]::arena::Arena>, &rustc_data_structures[e0d8483206daafa2]::sync::worker_local::WorkerLocal<rustc_hir[7eac7792dedc1861]::Arena>, rustc_driver_impl[517260a07841a630]::run_compiler::{closure#0}::{closure#2})>>::call_once::{shim:vtable#0}
  35:     0x7f3550bf5d34 - rustc_interface[5c48368f5debe80d]::interface::run_compiler::<(), rustc_driver_impl[517260a07841a630]::run_compiler::{closure#0}>::{closure#1}
  36:     0x7f3550b5aa3a - std[287527946f71f0bc]::sys::backtrace::__rust_begin_short_backtrace::<rustc_interface[5c48368f5debe80d]::util::run_in_thread_with_globals<rustc_interface[5c48368f5debe80d]::util::run_in_thread_pool_with_globals<rustc_interface[5c48368f5debe80d]::interface::run_compiler<(), rustc_driver_impl[517260a07841a630]::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}, ()>
  37:     0x7f3550b5a7fe - <std[287527946f71f0bc]::thread::lifecycle::spawn_unchecked<rustc_interface[5c48368f5debe80d]::util::run_in_thread_with_globals<rustc_interface[5c48368f5debe80d]::util::run_in_thread_pool_with_globals<rustc_interface[5c48368f5debe80d]::interface::run_compiler<(), rustc_driver_impl[517260a07841a630]::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}, ()>::{closure#1} as core[7477a31a1a34cc49]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  38:     0x7f3550b5c5c4 - <std[287527946f71f0bc]::sys::thread::unix::Thread>::new::thread_start
  39:     0x7f354a49698b - <unknown>
  40:     0x7f354a51a9cc - <unknown>
  41:                0x0 - <unknown>

error: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: please make sure that you have updated to the latest nightly

note: rustc 1.95.0-nightly (621d76794 2026-01-18) running on x86_64-unknown-linux-gnu

note: compiler flags: -Z next-solver=globally -Z dump-mir-dir=dir

query stack during panic:
#0 [supertrait_vtable_slot] finding the slot within vtable for trait object `dyn Supertrait<()>` vtable ptr during trait upcasting coercion from `dyn Trait<()>` vtable
#1 [eval_to_allocation_raw] const-evaluating + checking `_`
#2 [eval_to_const_value_raw] simplifying constant for the type system `_`
#3 [analysis] running analysis passes on crate `mvce`
end of query stack
error: aborting due to 2 previous errors

Some errors have detailed explanations: E0107, E0601.
For more information about an error, try `rustc --explain E0107`.

```

</p>
</details>

<!--
query stack:
#0 [supertrait_vtable_slot] finding the slot within vtable for trait object `dyn Supertrait<()>` vtable ptr during trait upcasting coercion from `dyn Trait<()>` vtable
#1 [eval_to_allocation_raw] const-evaluating + checking `_`
#2 [eval_to_const_value_raw] simplifying constant for the type system `_`
#3 [analysis] running analysis passes on crate `mvce`

-->
@rustbot label +WG-trait-system-refactor

```rust
#![feature(unboxed_closures)]
trait Supertrait<T> {}
trait Trait/*<T>*/: Supertrait<T> {}
impl Trait for () {}
const fn upcast<P>(x: &dyn Trait) -> &dyn Trait {
    x
}
const fn foo() -> &'static dyn Supertrait() {
    upcast::<()>(&())
}
const _: &'static dyn Supertrait<()> = foo();
fn main() {}
```

```
error[E0425]: cannot find type `T` in this scope
 --> code.rs:3:32
  |
3 | trait Trait/*<T>*/: Supertrait<T> {}
  |                                ^ not found in this scope

error[E0220]: associated type `Output` not found for `Supertrait`
 --> code.rs:8:32
  |
8 | const fn foo() -> &'static dyn Supertrait() {
  |                                ^^^^^^^^^^^^ associated type `Output` not found


thread 'rustc' (2106583) panicked at /rustc-dev/fe98ddcfcfb6f185dbf4adeaf439d8a756da0273/compiler/rustc_trait_selection/src/traits/vtable.rs:437:77:
called `Option::unwrap()` on a `None` value
stack backtrace:
....
```

not sure where this `Output` assoc type suggestion is coming from 🤔 


Regression in nightly-2025-02-01

```
  commit[0] 2025-01-30: Auto merge of #135030 - Flakebi:require-cpu, r=workingjubilee
  commit[1] 2025-01-30: Auto merge of #136318 - matthiaskrgr:rollup-a159mzo, r=matthiaskrgr
  commit[2] 2025-01-31: Auto merge of #135318 - compiler-errors:vtable-fixes, r=lcnr
  commit[3] 2025-01-31: Auto merge of #136331 - jhpratt:rollup-curo1f4, r=jhpratt
  commit[4] 2025-01-31: Auto merge of #136332 - jhpratt:rollup-aa69d0e, r=jhpratt
  commit[5] 2025-01-31: Auto merge of #134424 - 1c3t3a:null-checks, r=saethlin
  commit[6] 2025-01-31: Auto merge of #136350 - matthiaskrgr:rollup-6eqfyvh, r=matthiaskrgr
```

--- a/compiler/rustc_trait_selection/src/traits/vtable.rs
+++ b/compiler/rustc_trait_selection/src/traits/vtable.rs
@@ -434,7 +434,16 @@ pub(crate) fn supertrait_vtable_slot<'tcx>(
         }
     };
 
-    prepare_vtable_segments(tcx, source_principal, vtable_segment_callback).unwrap()
+    prepare_vtable_segments(tcx, source_principal, vtable_segment_callback).unwrap_or_else(|| {
+        // This can happen if the trait hierarchy is malformed (e.g., due to
+        // missing generics on a supertrait bound). There should already be an error
+        // emitted for this, so we just delay the ICE.
+        tcx.dcx().delayed_bug(format!(
+            "could not find the supertrait vtable slot for `{}` -> `{}`",
+            source, target
+        ));
+        None
+    })
 }
 
 pub(super) fn provide(providers: &mut Providers) {

--- a/tests/ui/traits/vtable/missing-generics-issue-151330.rs
+++ b/tests/ui/traits/vtable/missing-generics-issue-151330.rs
@@ -1,0 +1,23 @@
+//@ compile-flags: -Znext-solver=globally
+// Regression test for issue https://github.com/rust-lang/rust/issues/151330
+
+trait Supertrait<T> {}
+
+trait Trait<P>: Supertrait {}
+//~^ ERROR missing generics for trait `Supertrait`
+//~| ERROR missing generics for trait `Supertrait`
+//~| ERROR missing generics for trait `Supertrait`
+
+impl<P> Trait<P> for () {}
+
+const fn upcast<P>(x: &dyn Trait<P>) -> &dyn Trait<P> {
+    x
+}
+
+const fn foo() -> &'static dyn Supertrait<()> {
+    upcast::<()>(&())
+}
+
+const _: &'static dyn Supertrait<()> = foo();
+
+fn main() {}

--- a/tests/ui/traits/vtable/missing-generics-issue-151330.stderr
+++ b/tests/ui/traits/vtable/missing-generics-issue-151330.stderr
@@ -1,0 +1,53 @@
+error[E0107]: missing generics for trait `Supertrait`
+  --> $DIR/missing-generics-issue-151330.rs:6:17
+   |
+LL | trait Trait<P>: Supertrait {}
+   |                 ^^^^^^^^^^ expected 1 generic argument
+   |
+note: trait defined here, with 1 generic parameter: `T`
+  --> $DIR/missing-generics-issue-151330.rs:4:7
+   |
+LL | trait Supertrait<T> {}
+   |       ^^^^^^^^^^ -
+help: add missing generic argument
+   |
+LL | trait Trait<P>: Supertrait<T> {}
+   |                           +++
+
+error[E0107]: missing generics for trait `Supertrait`
+  --> $DIR/missing-generics-issue-151330.rs:6:17
+   |
+LL | trait Trait<P>: Supertrait {}
+   |                 ^^^^^^^^^^ expected 1 generic argument
+   |
+note: trait defined here, with 1 generic parameter: `T`
+  --> $DIR/missing-generics-issue-151330.rs:4:7
+   |
+LL | trait Supertrait<T> {}
+   |       ^^^^^^^^^^ -
+   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`
+help: add missing generic argument
+   |
+LL | trait Trait<P>: Supertrait<T> {}
+   |                           +++
+
+error[E0107]: missing generics for trait `Supertrait`
+  --> $DIR/missing-generics-issue-151330.rs:6:17
+   |
+LL | trait Trait<P>: Supertrait {}
+   |                 ^^^^^^^^^^ expected 1 generic argument
+   |
+note: trait defined here, with 1 generic parameter: `T`
+  --> $DIR/missing-generics-issue-151330.rs:4:7
+   |
+LL | trait Supertrait<T> {}
+   |       ^^^^^^^^^^ -
+   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`
+help: add missing generic argument
+   |
+LL | trait Trait<P>: Supertrait<T> {}
+   |                           +++
+
+error: aborting due to 3 previous errors
+
+For more information about this error, try `rustc --explain E0107`.

